cmd/ztest/ztest.c

```diff
@@ -6510,6 +6510,86 @@ ztest_run_zdb(char *pool)
 	umem_free(zbuf, 1024);
 }
 
+/*
+ * Use the zpool(8) command to find the config for 'pool'
+ */
+static nvlist_t *
+ztest_find_config(const char *pool, const char *devpath)
+{
+	nvlist_t *config = NULL;
+	struct stat64 statbuf;
+	char *bin;
+	char *zpool;
+	char *zbuf;
+	char *tmpfile;
+	char *packed;
+	const int len = MAXPATHLEN + MAXNAMELEN + 20;
+	FILE *fp;
+	int fd;
+
+	bin = umem_alloc(len, UMEM_NOFAIL);
+	zpool = umem_alloc(len, UMEM_NOFAIL);
+	zbuf = umem_alloc(1024, UMEM_NOFAIL);
+
+	ztest_get_binary("zpool", NULL, bin, len);
+
+	/*
+	 * Create a unique tempfile to hold the config
+	 */
+	tmpfile = umem_alloc(len, UMEM_NOFAIL);
+	(void) snprintf(tmpfile, len, "%s/ztest_pool_config_XXXXXX",
+	    ztest_opts.zo_dir);
+	fd = mkstemp(tmpfile);
+	ASSERT(fd > 0);
+	(void) close(fd);
+
+	(void) sprintf(zpool, "%s import -d %s -C %s %s",
+	    bin, devpath, tmpfile, pool);
+
+	if (ztest_opts.zo_verbose >= 5)
+		(void) printf("Executing %s\n", strstr(zpool, "zpool "));
+
+	fp = popen(zpool, "r");
+
+	while (fgets(zbuf, 1024, fp) != NULL)
+		if (ztest_opts.zo_verbose >= 3)
+			(void) printf("%s", zbuf);
+
+	if (pclose(fp) != 0) {
+		ztest_dump_core = 0;
+		(void) unlink(tmpfile);
+		(void) fatal(0, "No pools found\n");
+	}
+
+	umem_free(bin, len);
+	umem_free(zpool, len);
+	umem_free(zbuf, 1024);
+
+	fd = open(tmpfile, O_RDONLY);
+	if (fd < 0)
+		(void) fatal(0, "cannot open %s\n", tmpfile);
+	ASSERT0(fstat64(fd, &statbuf));
+	packed = umem_alloc(statbuf.st_size, UMEM_NOFAIL);
+
+	char *bufptr = packed;
+	int residual = statbuf.st_size;
+
+	while (residual > 0) {
+		ssize_t bytes = read(fd, bufptr, residual);
+		ASSERT(bytes >= 0);
+		bufptr += bytes;
+		residual -= bytes;
+	}
+	(void) close(fd);
+	(void) unlink(tmpfile);
+	umem_free(tmpfile, len);
+
+	config = fnvlist_unpack(packed, statbuf.st_size);
+	umem_free(packed, statbuf.st_size);
+
+	return (config);
+}
```

Comment:
This could go in `lib/libzpool/util.c` so it's available as part of `libzpool.so` but not the kmods.

Reply:
Good point.  I'll refactor into `lib/libzpool/util.c` and update `zdb.c` to also use it.